### PR TITLE
[SPARK-3910] Remove pyspark/mllib/ from sys.path in tests to fix relative import issue

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch addresses an issue where PySpark's MLlib unit tests might fail on certain environments due to circular import issues.  The root issue is that PySpark has a module named `random`, which shares its name as the built-in Python `random`.  This isn't a problem for consumers of Spark, though: `numpy` also has a `numpy.random` module, so it's perfectly fine to have _qualified_ module names that have components that match Python built-ins.  In normal operation, the top-level `pyspark` directory is added to `sys.path` so that PySpark's `random` module can only be imported via `pyspark.mllib.random`.  However, our unit tests end up invoking `./bin/pyspark python/pyspark/mllib/clustering.py`, which causes [the first entry of `sys.path`](https://docs.python.org/2/library/sys.html#sys.path) to be the directory containing `clustering.py` (`python/pyspark/mllib`, in that case); as a result, PySpark's `random` module shadows the built-in Python module.  This causes `numpy` and other dependencies to end up importing PySpark's `random`, which leads to circular import errors that cause tests to fail.

To fix this, we need to prevent our tests from adding the script's directory to `sys.path`.  There was already some code to do this in `linalg.py`, so this patch just copies that code to all of the other MLlib tests.

### CODE REVIEW

**Comment by JoshRosen** [comment]
TL;DR relative imports are a frustrating Python language feature :(

**Comment by JoshRosen** [comment]
Also, to be clear, I think this is only relevant for backports prior to Spark 1.2, which is why I've opened this against `branch-1.1`.

/cc @davies, I think this should be a safe patch because it only touches test code and because we don't use relative imports in our own code.

**Comment by JoshRosen** [comment]
Ah, I see why this failed: the `sys.path` hacking has to be done at the top of the file, not at the bottom.

**Comment by mengxr** [comment]
Would the hack we used in branch-1.2 and master work?

https://github.com/apache/spark/blob/master/python/pyspark/mllib/__init__.py

**Comment by davies** [comment]
@mengxr I think so, it's better to back port that into 1.1

**Comment by JoshRosen** [comment]
@davies @mengxr I agree with you; let's go ahead and backport the `master` and `branch-1.2` change, since I don't think it breaks compatibility from a user-POV and will make it easier to backport future fixes.
